1041/en.md

```diff
@@ -0,0 +1,160 @@
+
+### Code
+
+```
+#include <bits/stdc++.h>
+
+using namespace std;
+
+#define fast                          \
+    ios_base::sync_with_stdio(false); \
+    cin.tie(NULL);                    \
+    cout.tie(NULL);
+
+struct NewData
+{
+    string city1, city2;
+    int cost;
+} vr;
+
+bool operator<(struct NewData d1, struct NewData d2)
+{
+    return d1.cost < d2.cost;
+}
+
+vector<NewData> graph;
+map<string, bool> Checked;
+map<string, string> parent;
+set<string> st;
+map<string, set<string>> mp;
+
+void ClearAll()
+{
+    graph.clear();
+    Checked.clear();
+    parent.clear();
+    st.clear();
+    mp.clear();
+}
+
+string findParent(string s)
+{
+    if (parent[s] == s)
+        return s;
+    return findParent(parent[s]);
+}
+
+int MinCost()
+{
+    int res = 0;
+    for (auto x : st)
+    {
+        parent[x] = x;
+    }
+    for (auto x : graph)
+    {
+        string s = x.city1;
+        string s2 = x.city2;
+        string ps = findParent(s);
+        string ps2 = findParent(s2);
+        if (ps != ps2)
+        {
+            res += x.cost;
+            parent[ps2] = ps;
+        }
+    }
+    return res;
+}
+
+int bfs(string s)
+{
+    int res = 0;
+    queue<string> q;
+    q.push(s);
+    Checked[s] = true;
+    res++;
+    while (!q.empty())
+    {
+        string cur = q.front();
+        q.pop();
+        for (auto x : mp[cur])
+        {
+            if (!Checked[x])
+                Checked[x] = true, q.push(x), res++;
+        }
+    }
+    return res;
+}
+
+int main()
+{
+    int t, cs = 1;
+    cin >> t;
+    while (t--)
+    {
+        int n;
+        cin >> n;
+        ClearAll();
+        while (n--)
+        {
+            cin >> vr.city1 >> vr.city2 >> vr.cost;
+            graph.push_back(vr);
+            st.insert(vr.city1);
+            st.insert(vr.city2);
+            mp[vr.city1].insert(vr.city2);
+            mp[vr.city2].insert(vr.city1);
+        }
+        sort(graph.begin(), graph.end());
+        string ans = "";
+        int mnCost = MinCost();
+        ;
+        if (bfs(*st.begin()) != st.size())
+            ans = "Impossible\n";
+        cout << "Case " << cs++ << ": ";
+        if (ans.size() > 0)
+            cout << ans;
+        else
+            cout << mnCost << endl;
+    }
+    return 0;
+}
+```
+
+## Resources:
+- [BFS] (https://cp-algorithms.com/graph/breadth-first-search.html)
+- [DFS] (https://cp-algorithms.com/graph/depth-first-search.html) 
+- [Kruskal] (https://cp-algorithms.com/graph/mst_kruskal.html) 
+- [Prim's] (https://cp-algorithms.com/graph/mst_prim.html)
+
+
+### Tutorial by:
+Md Hasibur Rahman (Evan) <br/>
+Department of CSE, <br/>
+Jatiya Kabi Kazi Nazrul Islam University
```

Comment:
Add your LightOJ Profile link here instead, and keep these information in your profile page. Would be better organized.

Reply:
Done